### PR TITLE
refactor: decode conversation events - WPB-9520

### PR DESCRIPTION
**Pull request by johnxnguyen**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9520" title="WPB-9520" target="_blank">TaskWPB-9520</a>  [iOS] Decode conversation events
  </td></table>
  <br />
 

### Issue

Decode all conversation events, remove deprecated events.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

### CODE REVIEW

File: WireAPI/Sources/WireAPI/APIs/UpdateEventsAPI/UpdateEventType/ConversationEventType.swift [hunk 1]
```diff
@@ -20,21 +20,16 @@ import Foundation
...
     case delete = "conversation.delete"
-    case knock = "conversation.knock"
     case memberJoin = "conversation.member-join"
     case memberLeave = "conversation.member-leave"
     case memberUpdate = "conversation.member-update"
-    case messageAdd = "conversation.message-add"
     case messageTimerUpdate = "conversation.message-timer-update"
     case mlsMessageAdd = "conversation.mls-message-add"
     case mlsWelcome = "conversation.mls-welcome"
-    case otrAssetAdd = "conversation.otr-asset-add"
```
Comment (netbe) [inline WireAPI/Sources/WireAPI/APIs/UpdateEventsAPI/UpdateEventType/ConversationEventType.swift:37]: are these all types that aren't used anymore?

File: WireAPI/Sources/WireAPI/Models/Conversation/Conversation.swift [hunk 2]
```diff
@@ -18,25 +18,104 @@
...
+
+    /// How users can join a conversation.
+
+    public var access: Set<ConversationAccessMode>?
+
+    /// Which users are allowed to be participants.
+
+    public var accessRoles: Set<ConversationAccessRole>?
+
+    /// LEGACY: Which users are allowed to be participants.
```
Comment (netbe) [inline WireAPI/Sources/WireAPI/Models/Conversation/Conversation.swift:89]: we could add for which api version in comment

File: WireAPI/Sources/WireAPI/Models/Conversation/ConversationAccessMode.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+/// How users can join a conversation.
+
+public enum ConversationAccessMode: String, Codable {
+
+    /// Used in one-to-one/ and self conversations where no one else can
```
Comment (netbe) [inline WireAPI/Sources/WireAPI/Models/Conversation/ConversationAccessMode.swift]:
```suggestion
    /// Used in one-to-one and self conversations where no one else can
```

File: WireAPI/Tests/WireAPITests/APIs/UpdateEventsAPI/EventDecoding/ConversationEventDecodingTests.swift
Comment (netbe) [inline WireAPI/Tests/WireAPITests/APIs/UpdateEventsAPI/EventDecoding/ConversationEventDecodingTests.swift:1]: location of file should match the one in Sources.

Sources/WireAPI/UpdateEventsAPI/`Event decoding`
Tests/WireAPITests/APIS/UpdateEventsAPI/`EventDecoding`

**Review by netbe — APPROVED** [review]
LGTM;) minus a few comments

File: WireAPI/Sources/WireAPI/APIs/ConversationsAPI/ConversationsAPIV0.swift [hunk 4]
```diff
@@ -170,45 +170,45 @@ struct ConversationV0: Decodable, ToAPIModelConvertible {
...
-    var accessRoles: [String]?
+    var access: Set<ConversationAccessMode>?
+    var accessRoles: Set<ConversationAccessRole>?
     var creator: UUID?
     var epoch: UInt?
     var id: UUID?
     var lastEvent: String?
-    var lastEventTime: String?
-    var legacyAccessRole: String?
+    var lastEventTime: Date?
```
Comment (caldrian) [inline WireAPI/Sources/WireAPI/APIs/ConversationsAPI/ConversationsAPIV0.swift:179]: I had discovered an issue regarding timestamps a while ago.
Sometimes the backend sends timestamps with fractional seconds, sometimes without.
I wasn't able to have a single parser supporting both, so I added a workaround.
The ticket `WPB-6529` is mentioned in some TODO the code.
Please look at the comments in JIRA:
https://wearezeta.atlassian.net/browse/WPB-6529

File: WireAPI/Sources/WireAPI/APIs/UpdateEventsAPI/UpdateEventType/ConversationEventType.swift (hunk 1 above)
Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/APIs/UpdateEventsAPI/UpdateEventType/ConversationEventType.swift:37]: Correct, I couldn't find any usage of them in the iOS and Android codebases.

File: WireAPI/Sources/WireAPI/Models/Conversation/Conversation.swift (hunk 2 above)
Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/Models/Conversation/Conversation.swift:89]: done 604e1bdc7f8fcf3562430a65f3cba16ab3fc0975

File: WireAPI/Sources/WireAPI/Models/Conversation/ConversationAccessMode.swift (hunk 3 above)
Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/Models/Conversation/ConversationAccessMode.swift]: done 8d50f00455c65c591165e98d32c4dcc6c58f74a3

File: WireAPI/Tests/WireAPITests/APIs/UpdateEventsAPI/EventDecoding/ConversationEventDecodingTests.swift
Comment (johnxnguyen) [reply WireAPI/Tests/WireAPITests/APIs/UpdateEventsAPI/EventDecoding/ConversationEventDecodingTests.swift:1]: done fca0b272b4ea1aa5013d3c994004b5f08cb643b8

File: WireAPI/Sources/WireAPI/APIs/ConversationsAPI/ConversationsAPIV0.swift (hunk 4 above)
Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/APIs/ConversationsAPI/ConversationsAPIV0.swift:179]: hmm so it looks like backend just updated the documentation and change the format to be consistent. I'll take a second look and make sure we're decoding with the correct format. Thanks

Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/APIs/ConversationsAPI/ConversationsAPIV0.swift:179]: So it looks like that are payloads that have both time formats: for instance the `Client` payload has a `last_active` field that has time without milliseconds, and `time` with milliseconds. To handle this I can see two options:

1. Do manual decoding to handle decode the fields separately with different date decoders, or
2. Bring back the workaround that tries to decode with milliseconds then falls back to without milliseconds.

Option 2 seems fine to me, what do you think?

Comment (caldrian) [reply WireAPI/Sources/WireAPI/APIs/ConversationsAPI/ConversationsAPIV0.swift:179]: Option 2 sounds good to me.

Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/APIs/ConversationsAPI/ConversationsAPIV0.swift:179]: done 2143688ee4898b1a141123ff33edd625261a8376

File: WireAPI/Sources/WireAPI/Extensions/JSONDecoder+Default.swift [hunk 5]
```diff
@@ -29,14 +29,18 @@ extension JSONDecoder {
...
+            // same payload. To account for this, we try to decode both formats.
+            if let date = ISO8601DateFormatter.fractionalInternetDateTime.date(from: rawDate) {
+                return date
+            } else if let date = ISO8601DateFormatter.internetDateTime.date(from: rawDate) {
+                return date
+            } else {
                 throw DecodingError.dataCorruptedError(
                     in: container,
-                    debugDescription: "Expected date string to be ISO8601-formatted with fractional seconds"
+                    debugDescription: "Expected date string to be ISO8601-formatted with or without  fractional seconds"
```
Comment (caldrian) [inline WireAPI/Sources/WireAPI/Extensions/JSONDecoder+Default.swift]:
```suggestion
                    debugDescription: "Expected date string to be ISO8601-formatted with or without fractional seconds"
```

File: WireAPI/Sources/WireAPI/Extensions/JSONEncoder+Default.swift [hunk 6]
```diff
@@ -27,7 +27,7 @@ extension JSONEncoder {
         encoder.outputFormatting = .sortedKeys
         encoder.dateEncodingStrategy = .custom({ date, encoder in
             var container = encoder.singleValueContainer()
-            let transportString = ISO8601DateFormatter.default.string(from: date)
+            let transportString = ISO8601DateFormatter.fractionalInternetDateTime.string(from: date)
```
Comment (caldrian) [inline WireAPI/Sources/WireAPI/Extensions/JSONEncoder+Default.swift:30]: Is the backend actually able to handle both variants of timestamps, with and without fractional seconds, even if it sends different variants for certain fields?

**Review by caldrian — APPROVED** [review]
Thank you for the great effort!
From a bird's eye view it looks very well structured.
Since there have so many lines of code been added, I wasn't able to verify every line in detail.

File: WireAPI/Sources/WireAPI/Extensions/JSONEncoder+Default.swift (hunk 6 above)
Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/Extensions/JSONEncoder+Default.swift:30]: This I don't know, I think we can assume it's not, but current behavior is to always encode fractional seconds. In any case, I would open this discussion as a separate PR, if that's ok with you.

Comment (caldrian) [reply WireAPI/Sources/WireAPI/Extensions/JSONEncoder+Default.swift:30]: Of course.

**Pull request merged**